src/Tutor.Core/LearnerModel/DomainOverlay/KnowledgeComponentMasteries/SessionTracker.cs

```diff
@@ -0,0 +1,84 @@
+﻿using FluentResults;
+using System;
+using Tutor.Core.BuildingBlocks.EventSourcing;
+using Tutor.Core.LearnerModel.DomainOverlay.KnowledgeComponentMasteries.Events;
+using Tutor.Core.LearnerModel.DomainOverlay.KnowledgeComponentMasteries.Events.SessionLifecycleEvents;
+
+namespace Tutor.Core.LearnerModel.DomainOverlay.KnowledgeComponentMasteries
+{
+    public class SessionTracker : EventSourcedEntity
+    {
+        public TimeSpan DurationOfPreviousSessions { get; private set; } = new TimeSpan(0);
+        public TimeSpan DurationOfCurrentSession { get => HasActiveSession ? LastActivity.Value - StartOfCurrentSession.Value : new TimeSpan(0); }
+        public bool HasActiveSession { get => StartOfCurrentSession.HasValue; }
+        public DateTime? StartOfCurrentSession { get; private set; }
+        public DateTime? LastActivity { get; private set; }
+
+        public void Launch()
+        {
+            if (HasActiveSession) Causes(new SessionAbandoned());
+
+            Causes(new SessionLaunched());
+        }
+
+        public Result Terminate()
+        {
+            if (!HasActiveSession) return Result.Fail("No active session to terminate.");
+
+            Causes(new SessionTerminated());
+            return Result.Ok();
+        }
+
+        public Result Abandon()
+        {
+            if (!HasActiveSession) return Result.Fail("No active session to abandon.");
+
+            Causes(new SessionAbandoned());
+            return Result.Ok();
+        }
+
+        public override void Apply(DomainEvent @event)
+        {
+            When((dynamic)@event);
+        }
+
+        private void When(SessionLaunched @event)
+        {
+            StartOfCurrentSession = @event.TimeStamp;
+            LastActivity = @event.TimeStamp;
+
+            @event.DurationOfPreviousSessions = DurationOfPreviousSessions;
+            @event.DurationOfCurrentSession = DurationOfCurrentSession;
+        }
```

Comment:
I'm not sure if adding data to events inside the Apply method like I did here and in KCM is generally acceptable in event sourcing, but I feel like it's be preferable to trying to initialize the entire event when it's first instantiated in this specific case, since our events contain a lot of data which will be useful for analytics, but which isn't actually needed for applying changes to the aggregate. It ends up being a bit messy either way, but I feel like the behavior methods are a lot more readable this way.